### PR TITLE
Add copilot section to security page and add copilot

### PR DESCRIPTION
**Pull request by rushatgabhane**
### Fixed Issues

$ https://github.com/Expensify/App/issues/46925
PROPOSAL:

### Tests

Pre-req: enable newDotCopilot beta on your account.
Pre-req: From other accounts, add your account as a copilot

### Security page
1. Verify that you can see a new copilot section.
2. Note that the `three dot menu` won't be usable at this moment.
3. Verify that you can see `Add copilot` button.
4. Verify that you can see all accounts that you can copilot into.

### Add copilot
1. Add a copilot, select email and role.
2. Verify that you see a confirmation page.
3. Verify that you can add a copilot and it shows in security page.

### Offline add copilot
1. Verify that you can add a copilot while offline and it shows as greyed out in security page

- [x] Verify that no errors appear in the JS console

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/c2b8e36c-64ee-4ec0-bdc5-b0047eb3c876

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/2a896bd8-1205-4eb6-8a96-010182ba481e

https://github.com/user-attachments/assets/68c84ad0-460a-46b7-8827-3b2d3e92dd43

image

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Comment by shawnborton** [comment]
Design-wise this looks pretty good to me.

**Review by fedirjh — COMMENTED** [review]
@rushatgabhane when deeplinking to Add copilot screen, RHP modal should be opened within the security screen similar to other screens (2FA, close account)

https://github.com/user-attachments/assets/d8952616-300b-45ce-aa67-ccaeedc4b7af

**Comment by dangrous** [comment]
hi @rushatgabhane ! I was just informed - to deal with a fire, we're requiring a validation code for adding a delegate. I'm not entirely sure what that means for the UI / logic yet, I'm asking for an example of where we implemented it for a different flow.

We just started doing this, so I think there will probably be a recent PR we can copy

**Comment by dannymcclain** [comment]
@dangrous This is an example of what we're doing in a different flow to handle this:
image

So I imagine we could just fold this magic code RHP into the copilot flow wherever it makes sense. Something like this:
image

cc @Expensify/design

**Comment by shawnborton** [comment]
Yup, big fan of that pattern!

**Comment by dangrous** [comment]
neat, thank you! what's the issue number that's coming from? we'll also need to figure out how to actually send the code, as well, hopefully they've implemented that logic already and we can steal it :)

**Comment by rushatgabhane** [comment]
@dangrous so we'll have to update the API as well, right? send the 2fa code with `AddDelegate` command?

**Comment by dangrous** [comment]
yep, we'll send `validateCode` along with it.

Also @dannymcclain @shawnborton so the mocks you sent - that shows up after the user adds in the email they want to add. Like clicking "verify" sends the full request? or are you thinking it shows up at an earlier step?

**Comment by rushatgabhane** [comment]
i think verify should be the last step to avoid user fatigue?

**Comment by dannymcclain** [comment]
Yeah I'm not sure where I would expect it to be. I think I might expect it at the very beginning? Or maybe the very end. I don't know that I feel super strongly. @Expensify/design?

**Comment by shawnborton** [comment]
Same, I would expect it to be the very first thing.

**Comment by dangrous** [comment]
oh interesting, i was thinking put it at the end, so you'd only see it when you were ready to go

Selfishly, I think that might make it easier to build, also (right @rushatgabhane?) since we'd be sure of minimal delay between getting the code and sending it so it'd be sure to be valid still. Also we wouldn't have to worry about storing it in the meantime.

**Comment by rushatgabhane** [comment]
even if we don't store the code, we'll have to store related items like error, retry attempt etc

**Comment by rushatgabhane** [comment]
so I'd say we should do what we're doing in other places

**Comment by dannymcclain** [comment]
I guess I see where you're coming from, but I feel like I typically see these more at the beginning. Like "Hey, it looks like you're trying to make a serious change to your account, let's verify it's you before proceeding."

>so I'd say we should do what we're doing in other places of the app to be consistent

This also makes sense to me.

**Comment by dangrous** [comment]
great, yeah, let's follow the patterns we have if we can find examples of it. @rushatgabhane let us know if you have further questions

**Comment by rushatgabhane** [comment]
magic code is the last step - https://github.com/Expensify/App/pull/48320

**Comment by dangrous** [comment]
Okay great - let's copy that pattern here, using resendvalidatecode to verify the action. I think if we need to adjust later we can, but that seems like the best option for launch given that we need to get this merged asap.

Do you think you can take care of that today @rushatgabhane ?

**Comment by rushatgabhane** [comment]
> when deeplinking to Add copilot screen, RHP modal should be opened within the security screen similar to other screens (2FA, close account)


@fedirjh @DylanDylann do you know how to fix this one?

**Comment by rushatgabhane** [comment]
> do you know how to fix this one?

resolved

**Comment by rushatgabhane** [comment]
@fedirjh @DylanDylann this is ready for review

**Comment by rushatgabhane** [comment]
@fedirjh that is expected because backend has not been deployed yet.

**Comment by rushatgabhane** [comment]
the verify button is disabled when offline, so we're covered. what do you think?

**Comment by rushatgabhane** [comment]
good idea 👍

**Comment by shawnborton** [comment]
> It would be helpful if the button inside the confirmation page is disabled to avoid queuing validation code requests

Rather than disabling the button, shouldn't we go to a "You appear to be offline" page instead? I'm not quite sure when we actually should use disabled buttons for offline TBH. cc @Expensify/design for thoughts there.

**Comment by dannymcclain** [comment]
I _think_ disabling a button when offline is a pattern we have/is outlined in our offline patterns documentation. I think it would be fine to keep them here and disable the button with the offline indicator (IF that is something we do elsewhere). But I'm certainly not against what you're talking about Shawn.

**Comment by rushatgabhane** [comment]
yeah i just copied it from magic code page for adding new contact

**Comment by shawnborton** [comment]
Ah okay cool, if that's a correct pattern then let's go for it.

**Comment by shawnborton** [comment]
Any thoughts though @trjExpensify since you are an offline expert (strange for someone who is online all the time)

**Comment by trjExpensify** [comment]
Lololol. 

Pattern C (disabled button) is used when this is a form that you can otherwise interact with, but you can't submit the form when you're offline. Otherwise, we'd use a the full page blocking UI (Pattern D) if it isn't or we can't show important/accurate data on the page they're trying to access without being online. So I agree here we can use pattern C and disable the `Add copilot` button when offline. 👍

**Review by fedirjh — COMMENTED** [review]
Left some code changes. The magic code Form errors should be tested with the backend.

**Comment by rushatgabhane** [comment]
nice catch, i will have to set the error based on json code.
@dangrous what is the jsonCode when magic code is incorrect? For everything else, we can safely show the error from serverf?

**Comment by dangrous** [comment]
looks like it's `401 Not authorized - Invalid validateCode`

I think the other errors are fine to display to the user (without the error code if possible), we might want to polish that later but it's fine for now

**Comment by dangrous** [comment]
> Was playing around with the code and got this error `421 Delegates are not authorized to add delegates` , should this error be displayed to the user ?
> 
> Screenshot 2024-09-04 at 10 38 23 PM

Re: this one - were you trying to add a delegate from a delegate account? We should ideally not show the option to do that if you're logged in as a delegate. In theory the only thing you can do is remove yourself as a delegate.

**Comment by dangrous** [comment]
I think if we can? Not a huge priority but shouldn't be that hard I don't think, don't show the option to add if we have that delegate key in delegatedaccess

File: src/libs/actions/Delegate.ts
Comment (rushatgabhane) [reply src/libs/actions/Delegate.ts:234]: this is probably fine. we can refine it later if needed

File: src/pages/settings/Security/AddDelegate/DelegateMagicCodePage.tsx
Comment (rushatgabhane) [reply src/pages/settings/Security/AddDelegate/DelegateMagicCodePage.tsx:36]: good call

File: src/pages/settings/Security/AddDelegate/ValidateCodeForm/BaseValidateCodeForm.tsx
Comment (rushatgabhane) [reply src/pages/settings/Security/AddDelegate/ValidateCodeForm/BaseValidateCodeForm.tsx:65]: nice catch

File: src/pages/settings/Security/AddDelegate/ValidateCodeForm/BaseValidateCodeForm.tsx
Comment (rushatgabhane) [reply src/pages/settings/Security/AddDelegate/ValidateCodeForm/BaseValidateCodeForm.tsx:160]: this one is actually correct

File: src/pages/settings/Security/AddDelegate/ValidateCodeForm/BaseValidateCodeForm.tsx
Comment (rushatgabhane) [reply src/pages/settings/Security/AddDelegate/ValidateCodeForm/BaseValidateCodeForm.tsx:186]: nice catch

**Comment by rushatgabhane** [comment]
> Shoudn't be a space between the MenuItem and the description

no this one is as per figma and menuItem component.

**Comment by rushatgabhane** [comment]
> We should ideally not show the option to do that if you're logged in as a delegate. In theory the only thing you can do is remove yourself as a delegate.

done

**Comment by rushatgabhane** [comment]
@fedirjh all fixed and ready for a review

**Comment by rushatgabhane** [comment]
This is how everything looks so far. 

Ignore the loading at end because my account has been throttled because many requests)

https://github.com/user-attachments/assets/2e0e802a-6a7d-40cd-a5f0-b480c647bc82

**Review by fedirjh — COMMENTED** [review]
@rushatgabhane The magic form does not show errors if the validation code is invalid. The backend does not return any Onyx data in this case, and we may need to handle it on the frontend. I have suggested the required changes.

https://github.com/user-attachments/assets/2b759095-0ccb-4cab-acee-8e42a0b12e1f

---

When switching accounts, the delegate’s avatars are not rendered at the first time, you have to navigate to a different section and then return to see the avatars properly rendered.

https://github.com/user-attachments/assets/2f4ba800-2b55-49c8-8172-b222ee627551

**Comment by shawnborton** [comment]
Super nitpicky, but shouldn't there be SafeArea below the last card in this view on mobile? 
CleanShot 2024-09-06 at 15 12 56@2x

**Comment by shawnborton** [comment]
Looks like the email address is missing in the copy here? 
CleanShot 2024-09-06 at 15 13 48@2x

**Comment by rushatgabhane** [comment]
> When switching accounts, the delegate’s avatars are not rendered at the first time, you have to navigate to a different section and then return to see the avatars properly rendered.

unrelated to the PR

**Comment by rushatgabhane** [comment]
> Super nitpicky, but shouldn't there be SafeArea below the last card in this view on mobile?

that is how it is on other pages

**Comment by shawnborton** [comment]
Is it possible to add the SafeArea there?

**Comment by rushatgabhane** [comment]
> Looks like the email address is missing in the copy here?

My session had expired so it was missing the email

**Comment by rushatgabhane** [comment]
@shawnborton does this look better?

image

**Comment by shawnborton** [comment]
Ah no, the previous style looked better where the homebar overlaps the content - it's just that I think there would be extra padding at the very bottom of the view when you scroll to the bottom.

**Comment by rushatgabhane** [comment]
> it's just that I think there would be extra padding at the very bottom of the view when you scroll to the bottom

ahh got you!

**Comment by rushatgabhane** [comment]
is this better?

image

File: src/pages/settings/Security/SecuritySettingsPage.tsx
Comment (rushatgabhane) [reply src/pages/settings/Security/SecuritySettingsPage.tsx:69]: but we want to show the delegates optimistically

**Comment by shawnborton** [comment]
Hmm no, the space would be underneath the card that wraps the Copilots.

**Comment by rushatgabhane** [comment]
ahh okay you mean a margin

**Comment by rushatgabhane** [comment]
Alright, i increased the margin from 20px to 40px

image

**Comment by rushatgabhane** [comment]
Added `safeAreaBottomPadding`

image

**Comment by shawnborton** [comment]
Yes! That's it!

File: src/pages/settings/Security/AddDelegate/ValidateCodeForm/BaseValidateCodeForm.tsx
Comment (rushatgabhane) [reply src/pages/settings/Security/AddDelegate/ValidateCodeForm/BaseValidateCodeForm.tsx:130]: @fedirjh when we will dimiss the error, the modal will close because of the useEffect in `DelegateMagicCodePage` 

Do you have any ideas on how we can fix this?

Comment (rushatgabhane) [reply src/pages/settings/Security/AddDelegate/ValidateCodeForm/BaseValidateCodeForm.tsx]: im thinking we should add a pending field: `validateCode`

Comment (rushatgabhane) [reply src/pages/settings/Security/AddDelegate/ValidateCodeForm/BaseValidateCodeForm.tsx]: how about we add a pending field `validateCodePending`

it'll clear only for successData

Comment (rushatgabhane) [reply src/pages/settings/Security/AddDelegate/ValidateCodeForm/BaseValidateCodeForm.tsx]: that is a nice idea

**Comment by rushatgabhane** [comment]
@fedirjh can be merged after you approve. Does it look good now?

**Comment by rushatgabhane** [comment]
@fedirjh actually, i updated the existing delegate from the array. So there won't be duplicates now

File: src/pages/settings/Security/AddDelegate/ValidateCodeForm/BaseValidateCodeForm.tsx
Comment (rushatgabhane) [reply src/pages/settings/Security/AddDelegate/ValidateCodeForm/BaseValidateCodeForm.tsx]: Updated existing delegate from array to take care of dupes

https://github.com/Expensify/App/pull/47353#issuecomment-2334596009

**Comment by rushatgabhane** [comment]
Let's goooo!

**Comment by rushatgabhane** [comment]
Thank you for all the help, and consistent feedback @fedirjh

**Review by fedirjh — APPROVED** [review]
:shipit:

**Comment by rushatgabhane** [comment]
it hurts that `delegates` is an array. carried over from oldDot.

code would be cleaner if it was an object like every other onyx data

File: src/languages/es.ts
Comment (rushatgabhane) [inline src/languages/es.ts:5169]:
```suggestion
        makeSureItIsYou: "Vamos a asegurarnos de que eres tú",
```

File: src/components/MenuItemList.tsx
Comment (dangrous) [inline src/components/MenuItemList.tsx:67]: Is using `OfflineWithFeedback` here going to cause any weirdness in other menus?

**Review by dangrous — CHANGES_REQUESTED** [review]

File: src/pages/settings/Security/AddDelegate/AddDelegatePage.tsx
Comment (rushatgabhane) [reply src/pages/settings/Security/AddDelegate/AddDelegatePage.tsx:48]: this i do not control unfortunately

File: src/components/MenuItemList.tsx
Comment (rushatgabhane) [reply src/components/MenuItemList.tsx:67]: no because they won't be passing `children` will passed as they are

**Comment by rushatgabhane** [comment]
>  I can get that info too, have asked in Slack

appreciate it!

**Comment by rushatgabhane** [comment]
resolved

image

**Comment by rushatgabhane** [comment]
> I'm unable to add a copilot who's not already a contact/has an account

resolved

image

**Comment by rushatgabhane** [comment]
added learn more URL

**Comment by rushatgabhane** [comment]
@dangrous all done ✅

**Comment by dangrous** [comment]
Awesome, I will look shortly once my internet is better and I can play around more! @fedirjh can you give it another look as well?

**Comment by rushatgabhane** [comment]
it should be merged to 

```js
account {
delegatedAccess {
    delegates: [
        {
            email: abc@xyz.com,
            errorFields: {addDelegate: "error message"},
        }
    ]
}
}

```

**Comment by rushatgabhane** [comment]
but definitely shouldn't block this PR

**Comment by rushatgabhane** [comment]
@dangrous all yours!!

File: src/pages/settings/Security/AddDelegate/DelegateMagicCodePage.tsx
Comment (dangrous) [inline src/pages/settings/Security/AddDelegate/DelegateMagicCodePage.tsx:44]: We're so closeeeeee - I'm getting another issue with new accounts, those that aren't contacts already. The modal does not close after adding a fresh contact as a delegate. It works fine for existing contacts.

**Review by dangrous — CHANGES_REQUESTED** [review]
one outstanding issue with non-contact delegates, the modal doesn't close - but other than that we're good to go on my end! (And we can sort out any remaining bugs in the remaining PRs since it's still behind the beta.)

I will look into that backend error update, too, but agree that it shouldn't block.

File: src/pages/settings/Security/AddDelegate/DelegateMagicCodePage.tsx
Comment (rushatgabhane) [reply src/pages/settings/Security/AddDelegate/DelegateMagicCodePage.tsx:44]: interesting i had fixed it but let me try again

**Comment by rushatgabhane** [comment]
@dangrous resolved! i ditched `accountID` in url in favor of `login`.

**Review by dangrous — DISMISSED** [review]
let's do it!

**Review by dangrous — APPROVED** [review]

File: src/components/MenuItemList.tsx
Comment (CyberAndrii) [inline src/components/MenuItemList.tsx:62]: You forgot to move the `key` prop to this component so it prints an error in console now

Comment (rushatgabhane) [reply src/components/MenuItemList.tsx:62]: sorry i don't understand. OfflineWithFeedback should be passed `key` as a prop?

Comment (CyberAndrii) [reply src/components/MenuItemList.tsx:62]: Yup, since you wrapped `MenuItem` with `OfflineWithFeedback`, you have to move the `key` prop from `MenuItem` to `OfflineWithFeedback`

Screenshot 2024-09-18 at 22 58 37

Comment (rushatgabhane) [reply src/components/MenuItemList.tsx:62]: nice catch, let me send a PR

File: src/pages/settings/Security/AddDelegate/ConfirmDelegatePage.tsx
Comment (DylanDylann) [inline src/pages/settings/Security/AddDelegate/ConfirmDelegatePage.tsx:74]: This caused this [bug](https://github.com/Expensify/App/issues/52904). We need to replace the route instead of pushing the new route to the top of the stack.

File: src/pages/settings/Security/AddDelegate/SelectDelegateRolePage.tsx
Comment (eh2077) [inline src/pages/settings/Security/AddDelegate/SelectDelegateRolePage.tsx:59]: This caused 
- https://github.com/Expensify/App/issues/52903

The checkmark should be displayed when back to access level page for the access level selected, see https://github.com/Expensify/App/pull/55256

File: src/pages/settings/Security/AddDelegate/ConfirmDelegatePage.tsx
Comment (FitseTLT) [inline src/pages/settings/Security/AddDelegate/ConfirmDelegatePage.tsx:36]: This PR introduced personalDetails?.displayName ?? formatPhoneNumber(login) on the Copilot pages. For a phone-number account that never set a name, the backend stores the raw SMS login in displayName, so the truthy value wins the ?? and the formatting is skipped — surfacing +15551234567@expensify.sms in the UI. The fix (#97932) moves the formatting outside the ?? so it applies to whichever value is chosen.

**Pull request merged**
